### PR TITLE
bpo-42815: Fix issue when thread doesn't copy context of parent thread

### PR DESCRIPTION
**Pull request by uriyyo**
https://bugs.python.org/issue42815

### CODE REVIEW

**Comment by jcea** [comment]
Could you possibly explain why do you need to amend decimal tests?. I don't understand it.

**Comment by jcea** [comment]
Oh, I understand now. You want to preserve the behaviour of a thread messing with decimal context to affect other threads (including main thread).

I find this patch valuable, but the decimal context "interference" in current python could be something production code could depend on. In that situation, this patch is "risky" and we are possibly too late for 3.10, since 3.10rc1 is planned in a couple of days.

Maybe Python 3.11?

We also need a documentation update talking about the new "context" optional  parameter.

I requested a review in python-dev.

**Review by jcea — CHANGES_REQUESTED** [review]
1. Test should check that old context is altered when passing the old context as a parameter, and not altered when using an implicit new context.
2. The same when providing an explicit new context.

File: Lib/threading.py [hunk 1]
```diff
@@ -925,7 +930,7 @@ def run(self):
         """
         try:
             if self._target is not None:
-                self._target(*self._args, **self._kwargs)
+                self._context.run(self._target, *self._args, **self._kwargs)
```
Comment (vstinner) [inline Lib/threading.py:951]: That's one why to fix the issue. Another would be to modify the contextvars module to detect that it runs in a new thread. For exemple, this change doesn't cover the case of threads spawned by C libraries which then use the Python C API to run Python code. Or functions calling _thread.start_new_thread() directly.

I don't know which option is the best. At least, this change is reasonable small.

File: Lib/threading.py [hunk 2]
```diff
@@ -845,6 +846,10 @@ class is implemented.
         else:
             self._daemonic = current_thread().daemon
         self._ident = None
+        if context is not None:
+            self._context = context
+        else:
+            self._context = _copy_context()
```
Comment (vstinner) [inline Lib/threading.py:870]: Is _copy_context() fast/efficient when contextvars is not used at all?

File: Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Fix issue when new thread doesn't copy context of a parent thread. Patch
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst:1]: This description is not very helpful. I would prefer to rephrase it to explain that threads no longer **share** contextvars variables with their parent thread, but the context is now **copied**.

IMO this change is important enough to be documented in a ``.. versionchanged:: 3.11`` markup in https://docs.python.org/dev/library/threading.html#threading.Thread documentation.

**Review by vstinner — COMMENTED** [review]
cc @1st1, author of the contextvars module.

File: Lib/threading.py (hunk 2 above)
Comment (uriyyo) [reply Lib/threading.py:870]: This is information from [PEP-567](https://www.python.org/dev/peps/pep-0567/):
> The Context mapping is implemented using an immutable dictionary. This allows for a O(1) implementation of the copy_context() function. 

I believe it is fast/efficient to use `copy_context`.

File: Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst (hunk 3 above)
Comment (uriyyo) [reply Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst:1]: Actually every thread has it's own context and they do not share same context. This PR tries fix problem when child thread use empty context instead of copy of a parent context.

**Comment by uriyyo** [comment]
@jcea Actually every thread has it's own context and they do not share same context. This PR tries fix problem when child thread use empty context instead of copy of a parent context.

For instead, I can that this problem fixed at `asyncio.to_thread`:
https://github.com/python/cpython/blob/c7ea1e3dcea6fbc9842463ce2b785b43501b1eaa/Lib/asyncio/threads.py#L22-L25

File: Lib/threading.py (hunk 1 above)
Comment (uriyyo) [reply Lib/threading.py:951]: Sorry, maybe I add bad description for this issue. I will try to make things more clear. Every thread has it's own context and they do not share same context. This PR tries fix problem when child thread use empty context instead of copy of a parent context.

**Comment by jcea** [comment]
@uriyyo , your comment doesn't make sense if a thread modifying its context affects what other threads see. That mean "shared context" in my book :-).

The comment by @vstinner about threads launched by C extensions is interesting. I was thinking about detecting this case when getting the GIL and copy the context at that time. The problem would be to decide which context to copy, in the general case.

Maybe perfection is enemy of good and this PR would be good enough for now.

File: Lib/threading.py [hunk 4]
```diff
@@ -925,7 +930,7 @@ def run(self):
         """
         try:
             if self._target is not None:
-                self._target(*self._args, **self._kwargs)
+                self._context.run(self._target, *self._args, **self._kwargs)
         finally:
```
Comment (vstinner) [inline Lib/threading.py:952]: Please clear self._context in the finally block, as done for target.

Comment (uriyyo) [reply Lib/threading.py:952]: Fixed, thanks

**Comment by vstinner** [comment]
Currently, the `_contextvars` module is built as a dynamic library, it is not available when `setup.py build` is run by `make` to build Python, and so `import logging` fails when trying to get it indirectly.

I suggest to modify Modules/Setup to uncomment `#_contextvars (...)`, so the `_contextvars` module is built as a built-in module, and so is always available.

File: Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst (hunk 3 above)
Comment (vstinner) [reply Misc/NEWS.d/next/Library/2021-01-03-13-44-44.bpo-42815.IJqHPr.rst:1]: Please rephrase the NEWS entry and add the requested versionchanged markup.

**Comment by uriyyo** [comment]
Closing this PR as far as a related issue has been closed: https://github.com/python/cpython/issues/86981#issuecomment-1338522372
